Add Category model

ipantazi opened the pull request:
Added Category, CategoryController, CategoryServiceImpl classes.

Added CategoryDto, CreateCategoryRequestDto, BookDtoWithoutCategoryIds records and added categoryIds to CreateBookRequestDto and BookSearchParametersDto records. 

Added CategoryMapper, CategoryRepository, CategoryService interfaces.

Added categories to Book class, categoryIds to BookDto class, setCategoryIds, updateBookEntity and toBookDtoWithoutCategories methods to BookMapper interface. 

Added findAllByCategoryId method to BookRepository interface. 

Added getByCategoryId method to interface BookService and BookServiceImpl class. 

Added CategoryNotFoundException class and added handleAccessDeniedException, handleBadCredentials, handleGenericException, handleCategoryNotFoundException methods to CustomGlobalExceptionHandler class. 

Changes have been made to Liquibase.

ipantazi committed:
Added Category, CategoryController, CategoryServiceImpl classes.
Added CategoryDto, CreateCategoryRequestDto, BookDtoWithoutCategoryIds records and added categoryIds to CreateBookRequestDto and BookSearchParametersDto records.
Added CategoryMapper, CategoryRepository, CategoryService interfaces.
Added categories to Book class, categoryIds to BookDto class, setCategoryIds, updateBookEntity and toBookDtoWithoutCategories methods to BookMapper interface.
Added findAllByCategoryId method to BookRepository interface.
Added getByCategoryId method to interface BookService and BookServiceImpl class.
Added CategoryNotFoundException class and added handleAccessDeniedException, handleBadCredentials, handleGenericException, handleCategoryNotFoundException methods to CustomGlobalExceptionHandler class.
Changes have been made to Liquibase.

Elena-Bruyako commented on src/main/java/onlinebookstore/exception/CategoryNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package onlinebookstore.exception;
+
+public class CategoryNotFoundException extends RuntimeException {
```
remove

Elena-Bruyako commented on src/main/java/onlinebookstore/model/Book.java:
```diff
@@ -39,4 +47,13 @@ public class Book {
 
     private boolean isDeleted = false;
+
+    @ManyToMany(fetch = FetchType.LAZY)
```
```suggestion
    @ManyToMany
```

Elena-Bruyako commented on src/main/java/onlinebookstore/model/Category.java:
```diff
@@ -0,0 +1,32 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(nullable = false)
```
you can make it unique (optional)

Elena-Bruyako commented on src/main/java/onlinebookstore/model/User.java:
```diff
@@ -50,10 +50,9 @@ public class User implements UserDetails {
     private String shippingAddress;
 
-    @ManyToMany
+    @ManyToMany(fetch = FetchType.LAZY)
```
ManyToMany has `default FetchType.LAZY;`
```suggestion
    @ManyToMany
```

Elena-Bruyako commented on src/main/resources/db/changelog/changes/13-insert-caregories.yaml:
```diff
@@ -0,0 +1,35 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert-categories-to-categories-table
```
don't need to insert categories

Elena-Bruyako commented on src/main/resources/db/changelog/changes/14-add-is-deleted-to-categories.yaml:
```diff
@@ -0,0 +1,12 @@
+            tableName: categories
+            columns:
+              - column:
+                  name: is_deleted
```
replace to existing changelog 11

Elena-Bruyako commented on src/main/resources/db/changelog/changes/15-insert-books-part3.yaml:
```diff
@@ -0,0 +1,131 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert-books-to-books-table
```
don't need to insert books

Elena-Bruyako commented on src/main/resources/db/changelog/changes/16-assign-categories-for-existing-books.yaml:
```diff
@@ -0,0 +1,51 @@
+databaseChangeLog:
+  - changeSet:
+      id: assign-classic-categories-for-existing-books-to-books_categories-table
```
don't need to insert this data

Elena-Bruyako reviewed (changes requested):
see comments

ipantazi committed:
Deleted (fetch = FetchType.LAZY) from User and Book classes.
Added annotation unique = true to class Category.
Deleted CategoryNotFoundException from interface CategoryMapper, CustomGlobalExceptionHandler and BookServiceImpl.
Added refreshing information about category in categoriesCash in class CategoryServiceImpl.
Deleted insert-categories, add-is-deleted-column, insert-books and assign-classic-categories files, refresh db.changelog-master.yaml and 11-create-categories-table.yaml
